### PR TITLE
Validation of server identify in a CaaS environment #358

### PR DESCRIPTION
**Pull request by dz-1**

### CODE REVIEW

File: client/src/main/java/com/networknt/client/ssl/APINameChecker.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import javax.net.ssl.SSLException;
+
+import org.apache.http.conn.ssl.DefaultHostnameVerifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class APINameChecker {
+	private static final Logger logger = LoggerFactory.getLogger(APINameChecker.class);
+	private static final DefaultHostnameVerifier verifier = new DefaultHostnameVerifier();
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/APINameChecker.java]: Could you add some comments to this class & methods to describe what it does & how it's being used?
Also would there be benefit to unit test these to ensure validation criteria is maintained in the future?

File: client/src/main/java/com/networknt/client/ssl/APINameChecker.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	
+    public static boolean verify(final String name, final X509Certificate cert) {
+        try {
+        	verifier.verify(name, cert);
+            return true;
+        } catch (final SSLException ex) {
+            if (logger.isDebugEnabled()) {
+            	logger.debug(ex.getMessage(), ex);
+            }
+            return false;
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/APINameChecker.java]: Is debug ok for this or perhaps warn?

File: client/src/main/java/com/networknt/client/ssl/CientX509TrustManager.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import javax.net.ssl.X509TrustManager;
+
+/**
+ * Customized X509TrustManager. 
+ * Note: X509TrustManagers are eventually converted to X509ExtendedTrustManagers by sun.security.ssl.SSLContextImpl.AbstractTrustManagerWrapper
+ * 
+ * @author Daniel Zhao
+ *
+ */
+public class CientX509TrustManager implements X509TrustManager{
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/CientX509TrustManager.java]: I think some method level comments explaining what each of them do would go a long way

File: client/src/main/java/com/networknt/client/ssl/Light4jALPNClientSelector.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ *
+ */
+public class Light4jALPNClientSelector {
+	@SuppressWarnings("unchecked")
+	private static final Map<String, Object> TLS_CONFIG = (Map<String, Object>)Config.getInstance().getJsonMapConfig(Http2Client.CONFIG_NAME).get(Http2Client.TLS);
+	
+    public static void runAlpn(final SslConnection sslConnection, final ChannelListener<SslConnection> fallback, final ClientCallback<ClientConnection> failedListener, final ALPNProtocol... details) {
+        SslConduit conduit = UndertowXnioSsl.getSslConduit(sslConnection);
+        SSLEngine engine = UndertowXnioSsl.getSslEngine(sslConnection);
+
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/Light4jALPNClientSelector.java]: For me this method would definitely need some explanation

File: client/src/main/java/com/networknt/client/ssl/SSLUtils.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.networknt.utility.StringUtils;
+
+public class SSLUtils {
+	private static final Logger logger = LoggerFactory.getLogger(SSLUtils.class);
+	
+	public static Set<String> resolveTrustedNames(String trustedNames){
+		Set<String> nameSet = Arrays.stream(StringUtils.trimToEmpty(trustedNames).split(","))
+				.filter(StringUtils::isNotBlank)
+				.collect(Collectors.toSet());
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/SSLUtils.java]: Not sure this is the best approach for providing trusted names as it's on the application level rather then the connection level. Ie. for my service i have to set the list of all trusted names as part of all connections that i will be making, rather then setting the individual trusted names for a specific connection i'm creating. Thoughts?

File: client/src/main/java/com/networknt/client/ssl/X509TrustManagerDecorator.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.networknt.client.ssl;
+
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509ExtendedTrustManager;
+import javax.net.ssl.X509TrustManager;
+
+public class X509TrustManagerDecorator {
+	public static TrustManager[] decorate(TrustManager[] trustManagers, boolean checkIdentity, String trustedNames) {
+		if (null!=trustManagers && trustManagers.length>0) {
+			TrustManager[] decoratedTrustManagers = new TrustManager[trustManagers.length];
```
Comment (NicholasAzar) [inline client/src/main/java/com/networknt/client/ssl/X509TrustManagerDecorator.java]: What's this used for?

File: client/src/main/java/com/networknt/client/ssl/APINameChecker.java (hunk 1 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/APINameChecker.java]: ok. I'll add some unit tests. But this class is mostly a wrapper of the apache httpclient DefaultHostnameVerifier. Javadocs are available online. I'll add references to that class.

File: client/src/main/java/com/networknt/client/ssl/APINameChecker.java (hunk 2 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/APINameChecker.java]: will change the logging level.

File: client/src/main/java/com/networknt/client/ssl/CientX509TrustManager.java (hunk 3 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/CientX509TrustManager.java]: Method specifications are available in the javadoc of X509TrustManager. I'll add references to the interface.

File: client/src/main/java/com/networknt/client/ssl/Light4jALPNClientSelector.java (hunk 4 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/Light4jALPNClientSelector.java]: I'll add a short description.

File: client/src/main/java/com/networknt/client/ssl/SSLUtils.java (hunk 5 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/SSLUtils.java]: Do you mean you want make the trustedNames a constant once loaded?

File: client/src/main/java/com/networknt/client/ssl/X509TrustManagerDecorator.java (hunk 6 above)
Comment (dz-1) [reply client/src/main/java/com/networknt/client/ssl/X509TrustManagerDecorator.java]: This is used to convert X509TrustManagers loaded by service loaders to our customized ones.

**Comment by NicholasAzar** [comment]
@stevehu are you ok if a dependency to the apache http client is included as part of this?

**Comment by stevehu** [comment]
As of today, we don't have Apache HttpClient in our dependency but only in the test dependency for one of the modules. I have received a lot of issues for the library in the security front. Is there a reason we need it?

**Comment by dz-1** [comment]
I used 'DefaultHostnameVerifier' in it. Parsing 'common name' is a little bit complicated.
I can use 'BouncyCastle' implementation. I thought it's not necessary to introduce another dependency.

**Comment by stevehu** [comment]
Yes. We want to ensure that we only include dependencies when it is absolutely necessary. This will reduce the delivery size and make developers life easier without conflicts. If you only need one or two classes from another library, you can pick the source code and repackage it into our module. Remember to retain the copyright and authors etc. Thanks.

**Comment by dz-1** [comment]
I have removed the dependency from pom.xml.

**Comment by stevehu** [comment]
@dz-1 We should change the package to com.networknt, otherwise, users will have trouble to include http client in their application as there are duplicated classes in a fat jar. If the user's version is different then we are using, we don't even know which version is included in the final jar file. There are two options: change the package or change the class names. I would prefer the first option. What do you think?

**Comment by dz-1** [comment]
@stevehu I added a '.copied' suffix to all those packages.

File: client/src/main/java/com/networknt/client/ssl/SSLUtils.java (hunk 5 above)
Comment (NicholasAzar) [reply client/src/main/java/com/networknt/client/ssl/SSLUtils.java]: No I mean that if an api is initiating requests multiple services they will have to list the trusted names in the application config, instead as specifically for the individual connections being made. Does that make sense?

**Comment by NicholasAzar** [comment]
@stevehu FYI I had a discussion with @dz-1 yesterday and he's looking into the option of allowing per request trusted names, instead of a list of trusted names for all outbound connections from the service

**Comment by dz-1** [comment]
@NicholasAzar Changes have been checked in. Please review.

**Comment by NicholasAzar** [comment]
@dz-1 Yeah, working on it! Just need to spend a bit of time setting up a local test to ensure everything's still working

**Comment by NicholasAzar** [comment]
Verified mTLS still works (expected fail):
```
Statistics        Avg      Stdev        Max
  Reqs/sec       298.14     585.49   10628.82
  Latency      512.83ms   301.55ms      2.81s
  HTTP codes:
    1xx - 0, 2xx - 0, 3xx - 0, 4xx - 0, 5xx - 0
    others - 2479
  Errors:
    remote error: tls: bad certificate - 2479
  Throughput:     1.84MB/s
```
Verified happy path still works:
```
Statistics        Avg      Stdev        Max
  Reqs/sec     72647.93   10670.97   94776.09
  Latency        1.72ms     4.29ms   384.78ms
  HTTP codes:
    1xx - 0, 2xx - 726205, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:    18.52MB/s
```

MTLS Sample Timings:
```
    data_received..............: 270 MB  4.5 MB/s
    data_sent..................: 116 MB  1.9 MB/s
    http_req_blocked...........: avg=2.19µs   min=0s      med=0s       max=229.99ms p(90)=1µs      p(95)=1µs
    http_req_connecting........: avg=8ns      min=0s      med=0s       max=2.03ms   p(90)=0s       p(95)=0s
    http_req_duration..........: avg=233.32µs min=77µs    med=210µs    max=125.04ms p(90)=309µs    p(95)=360µs
    http_req_receiving.........: avg=17.11µs  min=3µs     med=13µs     max=75.06ms  p(90)=17µs     p(95)=25µs
    http_req_sending...........: avg=27.67µs  min=10µs    med=21µs     max=100.2ms  p(90)=25µs     p(95)=32µs
    http_req_tls_handshaking...: avg=1.66µs   min=0s      med=0s       max=228.48ms p(90)=0s       p(95)=0s
    http_req_waiting...........: avg=188.53µs min=0s      med=170µs    max=125ms    p(90)=262µs    p(95)=299µs
    http_reqs..................: 2703697 45061.313876/s
    iteration_duration.........: avg=306.21µs min=105.5µs med=268.71µs max=242.28ms p(90)=400.54µs p(95)=501.05µs
    iterations.................: 2703693 45061.247209/s
    vus........................: 20      min=20 max=20
    vus_max....................: 20      min=20 max=20
```

Verified one-way TLS still works:
```
Statistics        Avg      Stdev        Max
  Reqs/sec     71607.72   10502.07   96170.75
  Latency        1.74ms     4.52ms   419.07ms
  HTTP codes:
    1xx - 0, 2xx - 715896, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:    18.24MB/s
```

One-way TLS Sample Timings:
```
data_received..............: 274 MB  4.6 MB/s
    data_sent..................: 118 MB  2.0 MB/s
    http_req_blocked...........: avg=2.03µs   min=0s      med=0s       max=211.75ms p(90)=1µs      p(95)=1µs
    http_req_connecting........: avg=3ns      min=0s      med=0s       max=587µs    p(90)=0s       p(95)=0s
    http_req_duration..........: avg=225.74µs min=71µs    med=204µs    max=75.38ms  p(90)=300µs    p(95)=347µs
    http_req_receiving.........: avg=16.93µs  min=4µs     med=13µs     max=68.33ms  p(90)=17µs     p(95)=24µs
    http_req_sending...........: avg=27.36µs  min=10µs    med=21µs     max=75.37ms  p(90)=25µs     p(95)=31µs
    http_req_tls_handshaking...: avg=1.52µs   min=0s      med=0s       max=209.97ms p(90)=0s       p(95)=0s
    http_req_waiting...........: avg=181.44µs min=0s      med=165µs    max=68.49ms  p(90)=254µs    p(95)=289µs
    http_reqs..................: 2744219 45736.872701/s
    iteration_duration.........: avg=297.6µs  min=100.3µs med=262.89µs max=213.41ms p(90)=387.76µs p(95)=483.53µs
    iterations.................: 2744216 45736.822701/s
    vus........................: 20      min=20 max=20
    vus_max....................: 20      min=20 max=20
```

**Pull request merged**